tracks/ruby/exercises/high-scores/mentoring.md

```diff
@@ -0,0 +1,71 @@
+New concepts: Array query methods (+ chaining methods), instantiating a class, getter method or `attr_reader`.
+
+### Minimal solution for approval
+
+```ruby
+
+# version 2+
+
+class HighScores
+  def initialize(scores)
+    @scores = scores
+  end
+
+  def scores  # no need to change getter method to `attr_reader`
+    @scores
+  end
+
+  def personal_best
+    scores.max
+  end
+
+  def latest
+    scores.last
+  end
+
+  def personal_top
+    scores.sort.reverse.take(3) 
+  end
+
+  # composing a string either with string interpolation or with an if.. else.. statement
+  def report
+    difference = "#{personal_best - latest} short of" unless personal_best == latest
+
+    "Your latest score was #{latest}. That's #{difference} your personal best!".squeeze
+  end
+end
+
+```
+
+### Reasonable approaches
+
+- See the one above
+- Defining `@scores`, `@personal_best` and `personal_top` in the constructor.  
```

Comment:
Not sure I love this approach being here. It's very brittle and breaks as soon as the scores array is modified after instantiation (even though we don't test for this scenario). Would love to hear your thoughts!

Reply:
I just went with the fact that one of the other mentors approved this approach, to be honest. 
You're totally right that it would break after the scores are modified, but the thing is: we don't test that, and that's for a reason. And I don't like to reject solutions based on reasons that we don't test for. In pure TDD mode, this would suffice to get the tests to pass.
I would be very happy if there's another argument why this is not a good approach.

Reply:
> And I don't like to reject solutions based on reasons that we don't test for.

🤔 so do you always accept as soon as the tests pass? There's a lot of idiomatic nuance we can't realistically test for. This is a broader discussion, but from what I've read before I think a lot of mentors require more from a solution than just passing the tests, no?

For example, in Acronym I typically point to using `scan` when people use `split` as suggested in the mentor notes. Do you usually accept `split` solutions and then suggest `scan` or only accept when `scan` (or something in that spirit) is used?

That's why I'm mentioning this - I'd be ok with someone with more experience using this approach (after making sure they understand what it means),  but I'm afraid it can lead to confusion for novice programmers. So I would always follow an approach like this with a discussion on if and why it may be troublesome for you in the future, and I wouldn't recommend it for this exercise.

Reply:
I agree with @pgaspar , even though it's not tested for, a brittle solution is not a good guide for a mentor. If your example here is that "ok but TDD, this is barely enough", then why don't we accept shameless green solutions (solutions that basically have hardcoded solutions to the test cases)? **However**: I completely agree that solutions should _not_ be rejected if not covered by the instructions or test cases. Note here that IMO the instructions are part of the challenge and should not be disregarded. 

Example: A mentor on a different track mentored my `leap` exercise, suggesting I remove the parentheses, because their minimal solution did not have it in the guide. I had to explain that logically the solution to leap without parentheses is _not equivalent_.

Reply:
There's two questions in here:
1) Do we want brittle/non-idiomatic solutions to be approved, just because they pass the tests?
2) Is the 'everything in the initializer' an approvable approach?

1) No. 
2) I think not, I wouldn't do it myself, but why? If the _only_ argument against it is that it will break _if_ we'd add new _functionality_, I don't find that very convincing. 
If we think it's an approach that shouldn't be approved, or should be approved with food for thought, I'd like to add the arguments here that can help new mentors.
 
Four possible solutions in this matter, in the order of my liking:
1) Add a test that _proves_ this to be a brittle/bad solution.
2) An argument other than 'it will break if we add new functionality'. 
3) Add some explanation in the Mentor Notes here, that we can not _not_ approve this, due to the limitations of the exercise, so do `approve with food for thought` , including a recommendation to extract methods.
4) Add text to the instructions that implies that putting everything in the initializer is not an acceptable solution here.  

What do you all think?

Reply:
Hi Maud,

I think you're asking the right questions!! (sidenote: in GFMD, you have to add an extra newline if you want a new list, now it show's up like 1 2 3 4. Typing 1. 1. 1. 1. would have resulted in the same).

I actually think that the FOUR solutions you provide should _all_ / or at least the first three be implemented

> Add a test that proves this to be a brittle/bad solution.
If this is a possibility then yes. However, forcing `high scores` to be immutable, would also fix this issue. Testing that could be:
- init
- modify scores array outside of the class
- check if the outcome reflects the initial values

> An argument other than 'it will break if we add new functionality'.
My argument is the _Example_ I listed above. A "bad" minimal solution will have mentors use that as baseline.

> Is the 'everything in the initializer' an approvable approach?
Idiomatic ruby likes shallow but narrow classes, and plenty of methods, so I think you're right that it's not; however, calling methods to calculate the outcomes and memoize e.g. in instance variables from the initializer IMO would be perfectly fine.

Reply:
Hey Maud,

> I think not, I wouldn't do it myself, but why? If the only argument against it is that it will break if we'd add new functionality, I don't find that very convincing.
If we think it's an approach that shouldn't be approved, or should be approved with food for thought, I'd like to add the arguments here that can help new mentors.

My point is that we probably shouldn't even mention it.

Let's think of a different case: in the Minimal solution for approval we have in these notes replace `@scores` with `$scores`. It's totally valid Ruby and will pass all our tests. However, as people experienced with Ruby will notice this is not a Reasonable approach at all - yes it works for this specific case with the current tests but it will lead to problems as soon as you continue expanding the class or using it in slightly different ways, not described in the tests (just as the all in initializer approach does). However, probably most if not all attentive mentors would comment on the fact that `$scores` is a global variable and how that's a bad practice **in certain situations** not described in the specs (just as the all in initializer approach is).

Here's how it would break:

```ruby
a = HighScores.new([1, 2, 3])
b = HighScores.new([4, 5, 6])

p a.scores
# => [4,5,6]
p b.scores
# => [4,5,6]
```

Do I think we should add a note to these mentor notes saying that global variables are bad? No. Should we add a spec showing how this is a bad practice? Also no - this is just one of a long list of bad practices we would need to test against. I expect the mentors to point this out themselves. And if we added a note and/or a spec we'd have to add it to pretty much all exercises, because this is an issue in pretty much all of them.

The same thing happens for the all in initializer approach - I expect mentors to naturally comment on it, but I don't think we should call it a "reasonable" solution, nor do I think that we should add a note or a test in every exercise where using this approach passes the tests (there are a bunch of them!).

We may be running in circles a bit here. Maybe a comment from @kotp or @Insti could help? Thanks for your comments too @SleeplessByte :+1:

Reply:
> My point is that we probably shouldn't even mention it. <- Pedro

That makes perfect sense, actually! 

Let's go with that. 👍